### PR TITLE
Trigger cookie renewal on /_session

### PR DESCRIPTION
**Pull request by adrienverge**
It is currently not possible to programmatically renew an `AuthSession` cookie. The only way is to wait 10% of the cookie expiration time then make any request to CouchDB, so it sends a new cookie.

Programmatically getting cookies can be needed, for example, when passwords are managed by authentication server. E.g., a third-party auth server gets a cookie, forwards it to a user, and the user needs the CouchDB server to install a new cookie in his/her browser with a `Set-Cookie` header. (Browsers don't allow JS to install cookies, only a `Set-Cookie` header from the server can set the cookie.)

This patch brings the following change: an already-authenticated request (either POST or GET) on `/_session` will trigger the creation of a new cookie, whatever we're inside the 10% timeout window or not.  This way, applications can programmatically get a cookie without having to wait 60 seconds.

This closes: COUCHDB-3125

### CODE REVIEW

File: src/couch_httpd_auth.erl [hunk 1]
```diff
@@ -240,13 +242,21 @@ cookie_authentication_handler(#httpd{mochi_req=MochiReq}=Req, AuthModule) ->
         end
     end.
 
+cookie_auth_should_issue_new_cookie(MochiReq, TimeLeft, Timeout) ->
```
Comment (kxepal) [inline src/couch_httpd_auth.erl]: This is good name for test function, but not like this one. `maybe_issue_new_cookie` ?

Comment (adrienverge) [inline src/couch_httpd_auth.erl]: You're right!

Comment (adrienverge) [inline src/couch_httpd_auth.erl]: Thanks for you suggestion. I've updated the PR.

File: src/couch_httpd_auth.erl [hunk 2]
```diff
@@ -240,13 +242,21 @@ cookie_authentication_handler(#httpd{mochi_req=MochiReq}=Req, AuthModule) ->
         end
     end.
 
+maybe_issue_new_cookie(MochiReq, TimeLeft, Timeout) ->
```
Comment (rnewson) [inline src/couch_httpd_auth.erl]: should match on path_parts like elsewhere.

**Comment by rnewson** [comment]
not sure on the patch itself but please no one merge before 2.0.

**Comment by adrienverge** [comment]
Hey, thanks for reviewing.

> not sure on the patch itself

Do you mean not sure about the need for programmatical cookie renewal, or about the implementation? If there's a smarter/cleaner way of implementing this, I'll be glad to try!

**Comment by rnewson** [comment]
I think it's already a problem that a cookie can be extended indefinitely (sure, it's invalidated by a password change but I don't think everyone realises that they only expire through inactivity).

The patch itself is close to 'done', see my previous comments.

tbh I would rather not merge this, and also remove the 10% renewal thing as well; forcing users to re-authenticate at least once per couch_httpd_auth timeout period. That's a larger conversation, and a controversial one, and definitely a post 2.0 thing, but I think many administrators expect cookies to expire after that timeout regardless of activity.

I don't want to discourage you from contributing, you've done a great job in your first patch. Authentication is a gnarly corner and CouchDB's auth system is fiddly in its own right.

A way forward, post 2.0, is a conversation about sessions and particularly about permanent sessions. We might support new authentication methods, like token authentication, to deal with this situation.

**Comment by kxepal** [comment]
@adrienverge 
There will be more problem: authed user may eventually request `GET /_session` for the information about himself: username, roles etc. According the code, each time he'll do that, the new cookie will get issued. Sounds like not a good plan.

Isolating that behavior just for a some single method (like `POST`) could be good idea, but what @rnewson said, there need some discussion as the auth is very important thing and doesn't forgives any mistakes.

**Pull request closed without merging**